Don't unregister the TimerSet and timers if exception is thrown in Ti…

nemanja-boric-sociomantic opened the pull request:
…merExt

`TimerExt` is now swallowing all the unhandled exceptions, keeping the timers
registered by default. Previously, the uncaught exception would
propagate to TimerSet, unregistering it and making all TimerExt
registration not work again. That behaviour is now fixed, with the
default that, if exception is thrown, the timer will stay registered.

nemanja-boric-sociomantic committed:
Don't unregister the TimerSet and timers if exception is thrown in TimerExt

`TimerExt` is now swallowing all the unhandled exceptions, keeping the timers
registered by default. Previously, the uncaught exception would
propagate to TimerSet, unregistering it and making all TimerExt
registration not work again. That behaviour is now fixed, with the
default that, if exception is thrown, the timer will stay registered.

nemanja-boric-sociomantic commented:
Marked as a breaking change, because users may use `TimerExt` for purposes not covered by `DaemonApp`, so that could change the application's behavior.

gavin-norman-sociomantic commented:
LGTM, but will leave a bit in case anyone else has any comments.

mathias-lang-sociomantic commented:
Test ?

mathias-lang-sociomantic commented on src/ocean/util/app/ext/TimerExt.d:
```diff
@@ -27,6 +27,19 @@
+
+*******************************************************************************/
+
+static private Logger log;
```
Redundant static. Also last time a module-level logger was added it didn't go so well. Could you make it class-local?

nemanja-boric-sociomantic replied:
bah, I've copy/pasted it from somewhere else.

mathias-lang-sociomantic commented on src/ocean/util/app/ext/TimerExt.d:
```diff
@@ -27,6 +27,19 @@
 
 module ocean.util.app.ext.TimerExt;
 
+import ocean.util.log.Log;
```
`Logger`, `Log` will be gone in v4 and is deprecated in v3 but v4.x.x is too much behind

mathias-lang-sociomantic commented:
I suggest merging v3 into v4 (fast-forwarding?) before merging anything into it.

nemanja-boric-sociomantic commented:
> I suggest merging v3 into v4 (fast-forwarding?) before merging anything into it.

@mihails-strasuns-sociomantic said he's going to do this.

nemanja-boric-sociomantic commented:
Updated with the fix for the unittests, test for the catching behavior and `Log` -> `Logger` import.

mathias-lang-sociomantic commented on src/ocean/util/app/ext/TimerExt.d:
```diff
@@ -87,7 +87,7 @@ unittest
     }
 
-    new App;
+    (new App).run(null);
```
Not an unittest please, an integration test. `unittest` shouldn't do IO or affect global state, and `epoll` does.

nemanja-boric-sociomantic replied:
Well, there's no IO with this particular unittest. The `epoll_create` and `timerfd_create` calls already affect the global state, with one `epoll_wait` cycle or not.

mathias-lang-sociomantic replied:
Well, I think they shouldn't be called then :)

nemanja-boric-sociomantic replied:
Ok, I'll move the test into an integration test and remove `new App` from here.

nemanja-boric-sociomantic committed:
Don't instantiate `Application` inside unittests

Documented unittests shouldn't create sideeffect, which calling
epoll_create and timerfd_create certainly do.

nemanja-boric-sociomantic committed:
Add integration test for TimerExt

nemanja-boric-sociomantic commented:
Updated with removing instantantion of the applications inside unittests and moving the `TimerExt` test into `test/timerext/main.d`

nemanja-boric-sociomantic commented:
Replaced `e.toString()` with `getMsg(e)`, since the timers may be called multiple times per second, unlike in #227.

mathias-lang-sociomantic commented on test/timerext/main.d:
```diff
@@ -0,0 +1,89 @@
+    }
+}
+
+void main(istring[] args)
```
Missing space

nemanja-boric-sociomantic replied:
:grumpy-cat:

mathias-lang-sociomantic commented on test/timerext/main.d:
```diff
@@ -0,0 +1,89 @@
+               "console = false\n");
+
+    auto app = new App;
+    app.main(args);
```
Shouldn't you check the return code also ?

nemanja-boric-sociomantic commented:
updated: check the code of the `app.main` in test.

nemanja-boric-sociomantic commented:
I forgot the space >_>.

mihails-strasuns-sociomantic commented:
Please rebase

nemanja-boric-sociomantic commented:
Rebased.

mihails-strasuns-sociomantic commented:
Checked coverage stats manually and it seems correct - don't think we can do any better until reports from int tests are sent.
